Object detection pipeline

mishig25 opened the pull request:
# What does this PR do?
* Object detection pipeline
* Give an image or list of images, outputs obj det annotations in form:
```python
[
    [
        {'score': 0.9..., 'label': 'remote', 'box': [{'x': 66, 'y': 118}, ...},
    ],
    ...
]
```
* See [colab](https://colab.research.google.com/github/NielsRogge/Transformers-Tutorials/blob/master/DETR/DETR_minimal_example_(with_DetrFeatureExtractor).ipynb#scrollTo=3ynXL-OtGskG) for more details
<!--
Congratulations! You've made it this far! You're not quite done yet though.

Once merged, your PR is going to appear in the release notes with the title you set, so make sure it's a great title that fully reflects the extent of your awesome contribution.

Then, please replace this with a description of the change and which issue is fixed (if applicable). Please also include relevant motivation and context. List any dependencies (if any) that are required for this change.

Once you're done, someone will review your PR shortly (see the section "Who can review?" below to tag some potential reviewers). They may suggest changes to make the code even better. If no one reviewed your PR after a week has passed, don't hesitate to post a new comment @-mentioning the same persons---sometimes notifications get lost.
-->


## Before submitting
- [ ] This PR fixes a typo or improves the docs (you can dismiss the other checks if that's the case).
- [x] Did you read the [contributor guideline](https://github.com/huggingface/transformers/blob/master/CONTRIBUTING.md#start-contributing-pull-requests),
      Pull Request section?
- [x] Was this discussed/approved via a Github issue or the [forum](https://discuss.huggingface.co/)? Please add a link
      to it if that's the case. [huggingface_hub#74](huggingface/hub-docs#6)
- [ ] Did you make sure to update the documentation with your changes? Here are the
      [documentation guidelines](https://github.com/huggingface/transformers/tree/master/docs), and
      [here are tips on formatting docstrings](https://github.com/huggingface/transformers/tree/master/docs#writing-source-documentation).
- [x] Did you write any new necessary tests?


## Who can review?

Anyone in the community is free to review the PR once the tests have passed. Feel free to tag
members/contributors who may be interested in your PR.
@NielsRogge

<!-- Your PR will be replied to more quickly if you can figure out the right person to tag with @

 If you know how to use git blame, that is the easiest way, otherwise, here is a rough guide of **who to tag**.
 Please tag fewer than 3 people.
 

Models:

- albert, bert, xlm: @LysandreJik
- blenderbot, bart, marian, pegasus, encoderdecoder,  t5: @patrickvonplaten, @patil-suraj
- longformer, reformer, transfoxl, xlnet: @patrickvonplaten
- fsmt: @stas00
- funnel: @sgugger
- gpt2: @patrickvonplaten, @LysandreJik
- rag: @patrickvonplaten, @lhoestq
- tensorflow: @LysandreJik

Library:

- benchmarks: @patrickvonplaten
- deepspeed: @stas00
- ray/raytune: @richardliaw, @amogkam
- text generation: @patrickvonplaten
- tokenizers: @n1t0, @LysandreJik
- trainer: @sgugger
- pipelines: @LysandreJik

Documentation: @sgugger

HF projects:

- datasets: [different repo](https://github.com/huggingface/datasets)
- rust tokenizers: [different repo](https://github.com/huggingface/tokenizers)

Examples:

- maintained examples (not research project or legacy): @sgugger, @patil-suraj
- research_projects/bert-loses-patience: @JetRunner
- research_projects/distillation: @VictorSanh

 -->

mishig25 committed:
[WIP] Implement object-detection pipeline

mishig25 committed:
Define threshold const

mishig25 committed:
Add `threshold` argument

mishig25 committed:
Refactor

mishig25 committed:
Uncomment test inputs

mishig25 committed:
`rm

Co-authored-by: NielsRogge <48327001+NielsRogge@users.noreply.github.com>

NielsRogge commented on src/transformers/pipelines/object_detection.py:
```diff
@@ -0,0 +1,173 @@
+@add_end_docstrings(PIPELINE_INIT_ARGS)
+class ObjectDetectionPipeline(Pipeline):
+    """
+    Object detection pipeline using any :obj:`AutoModelForObjectDetection`. This pipeline predicts boundig boxes of
```
```suggestion
    Object detection pipeline using any :obj:`AutoModelForObjectDetection`. This pipeline predicts bounding boxes of
```

NielsRogge commented on src/transformers/pipelines/object_detection.py:
```diff
@@ -0,0 +1,173 @@
+
+    def __init__(
+        self,
+        model: Union["PreTrainedModel", "TFPreTrainedModel"],
```
Not sure if `TFPreTrainedModel` should already be included here.

mishig25 replied:
Removing "TFPreTrainedModel" for now then

NielsRogge commented on src/transformers/pipelines/object_detection.py:
```diff
@@ -0,0 +1,173 @@
+                The probability necessary to make a prediction.
+
+        Return:
+            A list of dictionaries or a list of of list of dictionaries containing result. If the input is a single
```
```suggestion
            A list of dictionaries or a list of list of dictionaries containing the result. If the input is a single
```

NielsRogge commented on src/transformers/pipelines/object_detection.py:
```diff
@@ -0,0 +1,173 @@
+
+            - **label** (:obj:`str`) -- The label identified by the model. There could be duplicates within the list
+              (corresponding to object of same label within an image).
+            - **score** (:obj:`int`) -- The score attributed by the model for that label.
```
Isn't the score a float rather than an int?

mishig25 replied:
It should be float (fixing it now)

NielsRogge commented on src/transformers/pipelines/object_detection.py:
```diff
@@ -0,0 +1,173 @@
+            outputs = self.model(**inputs)
+
+            target_sizes = torch.IntTensor([[im.height, im.width] for im in images])
+            annotations = self.feature_extractor.post_process(outputs, target_sizes)
```
The `post_process` method is specifically defined for DETR, not sure if other models need it. Perhaps we can make sure that each object detection model's feature extractor implements a `post_process` method, that converts the model outputs to the API we are expecting.

mishig25 replied:
I'd vote for 
> Perhaps we can make sure that each object detection model's feature extractor implements a post_process method, that converts the model outputs to the API we are expecting (which is the COCO detection API).

I'd like to get @Narsil suggestion

NielsRogge replied:
@Narsil can you provide your opinion here?

julien-c replied:
i think this sounds like a perfectly fine solution

Narsil replied:
I am not sure what my opinion on do you want ? On the output type, or the code ?

The code looks fine. You are introducing a necessity for every forward model to actually implement this `post_process` method. It really depends how much this code is model dependant.

In general, it's better to have generic code that works for every model, but it's not always possible (like `translation` or `conversational` where each models does it differently, in that case it's perfectly fine to delegate responsability to `tokenizer`/`feature_extractor`/`model` through a method.

Sorry about the very late review, notifications sometimes behaving for me, don't hesitate to ping on slack for these.

LysandreJik replied:
Also pinging @sgugger which might have an opinion regarding the `post_process` method to convert from prediction to understandable result.

I think having a `post_process` akin to the tokenizer's `decode` method is a good way to ensure that the output of a model/feature_extractor combo always remains understandable, so that approach is fine for me.

sgugger replied:
I also agree on requiring a `post_process` method for all object detection feature extractors.

NielsRogge commented on src/transformers/pipelines/object_detection.py:
```diff
@@ -0,0 +1,173 @@
+        Generates 4 corners of a box in a clockwise order (starting from top-left corner).
+
+        Args:
+            box (tensor): Tensor of shape (4) representng 2 opposing corners of object detection box
```
```suggestion
            box (tensor): Tensor containing the coordinates in corners format.
```

NielsRogge commented on src/transformers/pipelines/object_detection.py:
```diff
@@ -0,0 +1,173 @@
+
+            The dictionaries contain the following keys:
+
+            - **label** (:obj:`str`) -- The label identified by the model. There could be duplicates within the list
+              (corresponding to object of same label within an image).
```
```suggestion
            - **label** (:obj:`str`) -- The label identified by the model. 
```

NielsRogge commented on src/transformers/pipelines/object_detection.py:
```diff
@@ -0,0 +1,173 @@
+                annotation["boxes"] = [self._get_vertices(box) for box in boxes]
+
+                # {"scores": [...], ...} --> [{"score":x, ...}, ...]
+                plural2singular = {"scores": "score", "labels": "label", "boxes": "box"}
+                keys = [plural2singular[key] for key in annotation]
```
```suggestion
                keys = ["score", "label", "box"]
```

NielsRogge commented on src/transformers/pipelines/object_detection.py:
```diff
@@ -0,0 +1,173 @@
+
+        return annotations
+
+    def _get_vertices(self, box: Tensor) -> List[Dict[str, int]]:
```
```suggestion
    def _get_clockwise_vertices(self, box: Tensor) -> List[Dict[str, int]]:
```

NielsRogge commented on src/transformers/pipelines/object_detection.py:
```diff
@@ -0,0 +1,173 @@
+
+                annotation["scores"] = scores.tolist()
+                annotation["labels"] = [self.model.config.id2label[label.item()] for label in labels]
+                annotation["boxes"] = [self._get_vertices(box) for box in boxes]
```
```suggestion
                annotation["boxes"] = [self._get_clockwise_vertices(box) for box in boxes]
```

mishig25 committed:
Fix typo

Co-authored-by: NielsRogge <48327001+NielsRogge@users.noreply.github.com>

mishig25 committed:
Fix typo

Co-authored-by: NielsRogge <48327001+NielsRogge@users.noreply.github.com>

mishig25 committed:
Chore better doc

Co-authored-by: NielsRogge <48327001+NielsRogge@users.noreply.github.com>

mishig25 committed:
Rm unnecessary lines

Co-authored-by: NielsRogge <48327001+NielsRogge@users.noreply.github.com>

mishig25 committed:
Chore better naming

Co-authored-by: NielsRogge <48327001+NielsRogge@users.noreply.github.com>

mishig25 committed:
Update src/transformers/pipelines/object_detection.py

Co-authored-by: NielsRogge <48327001+NielsRogge@users.noreply.github.com>

mishig25 committed:
Update src/transformers/pipelines/object_detection.py

Co-authored-by: NielsRogge <48327001+NielsRogge@users.noreply.github.com>

mishig25 committed:
Fix typo

mishig25 committed:
Add `detr-tiny` for tests

LysandreJik commented on src/transformers/pipelines/__init__.py:
```diff
@@ -43,6 +43,7 @@
 from .feature_extraction import FeatureExtractionPipeline
 from .fill_mask import FillMaskPipeline
 from .image_classification import ImageClassificationPipeline
+from .object_detection import ObjectDetectionPipeline
```
This should also be added to the main `__init__` of `transformers`

LysandreJik commented on src/transformers/pipelines/object_detection.py:
```diff
@@ -0,0 +1,170 @@
+                The pipeline accepts either a single image or a batch of images, which must then be passed as a string.
+                Images in a batch must all be in the same format: all as http links, all as local paths, or all as PIL
+                images.
+            threshold (:obj:`float`, `optional`, defaults to 0.9):
+                The probability necessary to make a prediction.
```
Nice, useful input!

LysandreJik commented on tests/test_pipelines_object_detection.py:
```diff
@@ -0,0 +1,227 @@
+@require_timm
+@require_vision
+@require_torch
+class ObjectDetectionPipelineTests(unittest.TestCase):
```
Thanks for implementing a nice set of tests! Unfortunately, we went through a refactor with @Narsil that will require updating this file to conform to the new testing format.

LysandreJik replied:
You can take a look at the new file for testing the image classification pipeline to get an idea of what needs to be done: https://github.com/huggingface/transformers/blob/master/tests/test_pipelines_image_classification.py

Please let us know if you'd like us to handle this refactor.

mishig25 replied:
awesome! Yeah, I'd like to handle this refactor

LysandreJik reviewed (commented):
Thanks a lot for proposing this PR @mishig25!

One remark, that is very possibly linked to the underlying model (DETR) rather than the pipeline itself: the pipeline doesn't seem to work with images with 4 dimensions (such as an alpha channel): 
```
ValueError: operands could not be broadcast together with shapes (4,1331,395) (3,) 
```
Attaching an image here to try it out with:
![rocket](https://user-images.githubusercontent.com/30755778/131346476-af2b7b88-c107-40d0-97cb-4a9cf22d4854.png)

Otherwise LGTM. Since this PR is bound to induce merge conflicts with https://github.com/huggingface/transformers/pull/13308, which one would you prefer we merge first @Narsil?

sgugger reviewed (approved):
Thanks a lot for adding this new pipeline!

Narsil commented on src/transformers/pipelines/object_detection.py:
```diff
@@ -0,0 +1,170 @@
+        with torch.no_grad():
+            inputs = self.feature_extractor(images=images, return_tensors="pt")
+            outputs = self.model(**inputs)
+
+            target_sizes = torch.IntTensor([[im.height, im.width] for im in images])
```
This is Pytorch specific, it should at the very least be protected by `if self.framework == "pt"`.

I know there's already a guard in the __init__, but pipelines should avoid as much as possible framework specific code (got bitten by this in ASR when TF was added later)

mishig25 replied:
Made the suggested change 👍

Narsil commented on src/transformers/pipelines/object_detection.py:
```diff
@@ -0,0 +1,170 @@
+
+                annotation["scores"] = scores.tolist()
+                annotation["labels"] = [self.model.config.id2label[label.item()] for label in labels]
+                annotation["boxes"] = [self._get_clockwise_vertices(box) for box in boxes]
```
I am not a big fan of this format.

bounding boxes are quite standarly defined by either:

- x, y, width, height
- x_min, x_max, y_min, y_max

A list of (x, y) is both uncommon and too general (we can't be sure it's a rectangle from a user's perspective so drawing anything will require additional computing)

mishig25 replied:
Implementing `x_min, x_max, y_min, y_max` format in the update 👍

Narsil commented on src/transformers/pipelines/object_detection.py:
```diff
@@ -0,0 +1,170 @@
+
+                # {"scores": [...], ...} --> [{"score":x, ...}, ...]
+                keys = ["score", "label", "box"]
+                annotation = [dict(zip(keys, vals)) for vals in zip(*annotation.values())]
```
You cannot do this this way.
Dict in python are not of deterministic order so `annotation.values()` is not guaranteed to be in the correct order ('score', 'label', 'box').
Even worse, this dependency is not explicity for the reader (so anyone might want to reorder the `keys` list and incur a bug.

```python
n = len(annotations["scores"]) # We're hopeful that annotations are actually of sane outputs with the same length of every key, that's very hopeful.

```python
annotation = [{"score": raw_annotation["score"][i], "label": raw_annotation["label"][i], "box": raw_annotation["box"][i] for i in range(n)]
```
I would argue that the real culprit is the `post_process` tbh, because the output might be *unsound* (different shapes for annotations is possible). As much as possible, we want to use struct/data formats that cannot allow any discrepancy to even exist (list of Annotation, won't allow any different length, which woudl result in erroneous attribution)

Let's see if we can move this logic directly within `post_process`. It seems much more natural in terms of responsability but again, I don't know who else uses `post_process` out there, so my assumptions might be wrong.

mishig25 replied:
I'll incorporate the changes 👍 
I've banked on the fact that since py3.6 dicts are insertion order preserving [here](https://docs.python.org/3/whatsnew/3.6.html#whatsnew36-compactdict). But even the doc says: 
> The order-preserving aspect of this new implementation is considered an implementation detail and **should not be relied upon**

Narsil reviewed (commented):
Thank you for writing this ! 
A few remarks here and there about the code.

The most important point is list of (x, y) vs actual bounding box, which I think the latter is preferable.

The rest is more about code hygiene and python caveats.

Narsil commented:
@LysandreJik ofc as a personal matter I would rather this would be merged after the large refactor, but tbh this should be handled like any other PR, the first to be ready should be approved.

Moving this code to the new PR should be just as easy as the other ones (the code is currently rather well separated in terms of concerns). The biggest concern regarding code separation is one I raise here which is the scope of `post_process`. I would advocate it should take on more (so the pipelines delegates ALL work to the model) but it might be difficult for various reasons I don't measure.

mishig25 committed:
Add `ObjectDetectionPipeline` to `trnsfrmrs/init`

mishig25 committed:
Implement new bbox format

mishig25 committed:
Merge branch 'master' into object_detection_pipeline

mishig25 committed:
Update detr post_process

mishig25 committed:
Update `load_img` method obj det pipeline

mishig25 committed:
make style

mishig25 commented:
@NielsRogge could you confirm that Detr model input can handle 4channel images as notified [here](https://github.com/huggingface/transformers/pull/12886#pullrequestreview-741650464)?

mishig25 committed:
Implement new testing format for obj det pipeln

mishig25 committed:
Add guard pytorch specific code in pipeline

mishig25 committed:
Add doc

mishig25 commented:
Re-requesting review from @Narsil and @LysandreJik 

@Narsil,
1. Changed box format to be {xmin,ymin,xmax,ymax} [here](https://github.com/huggingface/transformers/blob/de23a8cfca50f5f166793fdd9d31458bf94f360d/src/transformers/pipelines/object_detection.py#L147-L164)
2. Added `self.framework == pt` guard on pytorch specific code [here](https://github.com/huggingface/transformers/blob/de23a8cfca50f5f166793fdd9d31458bf94f360d/src/transformers/pipelines/object_detection.py#L147-L164)
3. As suggested by comment [here](https://github.com/huggingface/transformers/pull/12886#discussion_r698575417), `post_process` is handling more responsibility. As a side effect, [this shadowing](https://github.com/huggingface/transformers/pull/12886#discussion_r698567466) concern disappears

@LysandreJik 
1. RGBA images are being handled when I updated `load_image` method (copied updates from image classification) [here](https://github.com/huggingface/transformers/blob/de23a8cfca50f5f166793fdd9d31458bf94f360d/src/transformers/pipelines/object_detection.py#L64-L83)
2. Added `ObjectDetectionPipeline` to [transformers/__init__.py](https://github.com/huggingface/transformers/blob/3f22f6d8393bd20dae5f875ec39f2adbd33d1d33/src/transformers/__init__.py)
3. Updated the testing file to match with updated testing scheme [here](https://github.com/huggingface/transformers/blob/4a8449ee18506d749da3291ac1df4b5dfefd8f62/tests/test_pipelines_object_detection.py)

Please let me know if you encounter any questions or concerns 👍

Narsil commented on tests/test_pipelines_object_detection.py:
```diff
@@ -0,0 +1,263 @@
+        outputs = object_detector("http://images.cocodataset.org/val2017/000000039769.jpg")
+
+        self.assertGreater(len(outputs), 0)
+        for detected_object in outputs:
+            self.assertEqual(
+                detected_object,
+                {
+                    "score": ANY(float),
+                    "label": ANY(str),
+                    "box": {"xmin": ANY(int), "ymin": ANY(int), "xmax": ANY(int), "ymax": ANY(int)},
+                },
+            )
+
```
No this test is deterministic, we need to enforce that the values are explicit.

The goal of `test_small_model` and `test_large_model` are to check that values do not drift over time (some changes in the pipelines might cause some drift)

mishig25 replied:
what is meant by `deterministic`: the length of `outputs`?

Narsil commented on tests/test_pipelines_object_detection.py:
```diff
@@ -0,0 +1,263 @@
+        for outputs in batch_outputs:
+            self.assertGreater(len(outputs), 0)
+            for detected_object in outputs:
+                self.assertEqual(
+                    detected_object,
+                    {
+                        "score": ANY(float),
+                        "label": ANY(str),
+                        "box": {"xmin": ANY(int), "ymin": ANY(int), "xmax": ANY(int), "ymax": ANY(int)},
+                    },
+                )
```
Same here.

Narsil commented on tests/test_pipelines_object_detection.py:
```diff
@@ -0,0 +1,263 @@
+                )
+
+    @require_torch
+    @slow
+    def test_ntegration_torch_object_detection(self):
+        model_id = "facebook/detr-resnet-50"
+
+        object_detector = pipeline("object-detection", model=model_id)
+
+        outputs = object_detector("http://images.cocodataset.org/val2017/000000039769.jpg")
+
+        self.assertGreater(len(outputs), 0)
+        for detected_object in outputs:
+            self.assertEqual(
+                detected_object,
+                {
+                    "score": ANY(float),
+                    "label": ANY(str),
+                    "box": {"xmin": ANY(int), "ymin": ANY(int), "xmax": ANY(int), "ymax": ANY(int)},
+                },
+            )
+
+        batch = [
+            "http://images.cocodataset.org/val2017/000000039769.jpg",
+            "http://images.cocodataset.org/val2017/000000039769.jpg",
+        ]
+        batch_outputs = object_detector(batch)
+
+        self.assertEqual(len(batch), len(batch_outputs))
+        for outputs in batch_outputs:
+            self.assertGreater(len(outputs), 0)
+            for detected_object in outputs:
+                self.assertEqual(
+                    detected_object,
+                    {
+                        "score": ANY(float),
+                        "label": ANY(str),
+                        "box": {"xmin": ANY(int), "ymin": ANY(int), "xmax": ANY(int), "ymax": ANY(int)},
+                    },
```
This test seems to be redundant with `test_large_model_pt`

Narsil commented on tests/test_pipelines_object_detection.py:
```diff
@@ -0,0 +1,263 @@
+                    },
+                )
+
+    def test_custom_tokenizer(self):
+        model_id = "facebook/detr-resnet-50"
+        tokenizer = PreTrainedTokenizer()
+
+        # Assert that the pipeline can be initialized with a feature extractor that is not in any mapping
+        object_detector = pipeline("object-detection", model=model_id, tokenizer=tokenizer)
+
+        self.assertIs(object_detector.tokenizer, tokenizer)
```
Not sure we need this one.

It is in `image-classification` but I am not sure we nede to copy it.

Narsil commented on tests/test_pipelines_object_detection.py:
```diff
@@ -0,0 +1,263 @@
+
+        self.assertIs(object_detector.tokenizer, tokenizer)
+
+    def test_low_threshold(self):
+        threshold = 0.0
+        model_id = "mishig/tiny-detr-mobilenetsv3"
+
+        model = AutoModelForObjectDetection.from_pretrained(model_id)
+        feature_extractor = AutoFeatureExtractor.from_pretrained(model_id)
+        object_detector = ObjectDetectionPipeline(model=model, feature_extractor=feature_extractor)
```
This needs to be `@require_torch`.

Narsil commented on tests/test_pipelines_object_detection.py:
```diff
@@ -0,0 +1,263 @@
+        self.assertTrue(isinstance(output, list))
+        self.assertEqual(len(output), 5)
+
+    def test_high_threshold(self):
+        threshold = 1.0
+        model_id = "mishig/tiny-detr-mobilenetsv3"
+
+        model = AutoModelForObjectDetection.from_pretrained(model_id)
+        feature_extractor = AutoFeatureExtractor.from_pretrained(model_id)
+        object_detector = ObjectDetectionPipeline(model=model, feature_extractor=feature_extractor)
+
+        valid_input = {"images": "http://images.cocodataset.org/val2017/000000039769.jpg", "threshold": threshold}
+
+        output = object_detector(**valid_input)
+
+        self.assertTrue(isinstance(output, list))
+        self.assertEqual(len(output), 0)
```
I think we can fuse `low_threshold` and `high_threshold` into 1 test, which would be testing a bit more specifically.

Currently the test would be failing on a lot of changes in the threshold code (you're only testing the extreme cases).

I think a better test would be to first get all top_k = 5 first results, take a threshold between 2nd and 3rd options (supposing they are different).

Because this test is deterministic we can even specify a hardcoded value for that threshold.
IMO that would be a better in checking that the filtering is filtering exactly what we expect it to.

Narsil commented on tests/test_pipelines_object_detection.py:
```diff
@@ -0,0 +1,263 @@
+
+    @require_torch
+    def test_small_model_pt(self):
+        threshold = 0.0
+        model_id = "mishig/tiny-detr-mobilenetsv3"
+
+        model = AutoModelForObjectDetection.from_pretrained(model_id)
+        feature_extractor = AutoFeatureExtractor.from_pretrained(model_id)
+        object_detector = ObjectDetectionPipeline(model=model, feature_extractor=feature_extractor)
+
+        outputs = object_detector("http://images.cocodataset.org/val2017/000000039769.jpg", threshold=threshold)
+
+        self.assertGreater(len(outputs), 0)
+        for detected_object in outputs:
+            self.assertEqual(
+                detected_object,
+                {
+                    "score": ANY(float),
+                    "label": ANY(str),
+                    "box": {"xmin": ANY(int), "ymin": ANY(int), "xmax": ANY(int), "ymax": ANY(int)},
+                },
+            )
+
+        batch = [
```
Same here, this needs to be deterministic.

Narsil commented on src/transformers/models/detr/feature_extraction_detr.py:
```diff
@@ -673,7 +673,7 @@ def pad_and_create_pixel_mask(
-    def post_process(self, outputs, target_sizes):
+    def post_process(self, outputs, target_sizes, threshold=0.9):
         """
         Converts the output of :class:`~transformers.DetrForObjectDetection` into the format expected by the COCO api.
         Only supports PyTorch.
```
This part seems pretty important (COCO api format). Are we sure we're not breaking anything ?

This is what I feared when I mentioned it might be hard to change this.

Narsil reviewed (commented):
Seems much better.

the `post_process` part seems like we're breaking something :(

Also added some comments to try and improve the tests.

mishig25 committed:
Make pipeline_obj_tet tests deterministic

mishig25 committed:
Revert some changes to `post_process` COCO api

mishig25 committed:
Chore

Narsil reviewed (approved):
LGTM.

Thanks !

NielsRogge commented on src/transformers/pipelines/object_detection.py:
```diff
@@ -0,0 +1,178 @@
+        if self.framework == "tf":
+            raise ValueError(f"The {self.__class__} is only available in PyTorch.")
+
+        requires_backends(self, "timm")
```
This is probably only a requirement for DETR. Other object detection models in the future might not rely on timm.

mishig25 replied:
@NielsRogge what would you suggest we do here?
add DETR specific check? `if model == detr, then requires_backends(self, "timm")`?

NielsRogge replied:
Hmm but users should pass a model themselves at initialization? Then that check is already done?

Or is this to load the default object detection model of the pipeline?

mishig25 replied:
Currently, the default model is detr. But as you've said:
> model themselves at initialization? Then that check is already done?
Therefore, I should just remove `requires_backends(self, "timm")`. Right?

Basically, `requires_backends(self, "timm")` should check timm on default model initialization, which is detr.

NielsRogge commented on src/transformers/pipelines/object_detection.py:
```diff
@@ -0,0 +1,178 @@
+                    f"Incorrect path or url, URLs must start with `http://` or `https://`, and {image} is not a valid path"
+                )
+        elif isinstance(image, Image.Image):
+            image = image
```
```suggestion
            pass
```

NielsRogge commented on src/transformers/pipelines/object_detection.py:
```diff
@@ -0,0 +1,178 @@
+            image = image
+        else:
+            raise ValueError(
+                "Incorrect format used for image. Should be an url linking to an image, a local path, or a PIL image."
```
```suggestion
                "Incorrect format used for image. Should be a URL linking to an image, a local path, or a PIL image."
```

NielsRogge commented on src/transformers/pipelines/object_detection.py:
```diff
@@ -0,0 +1,178 @@
+            images (:obj:`str`, :obj:`List[str]`, :obj:`PIL.Image` or :obj:`List[PIL.Image]`):
+                The pipeline handles three types of images:
+
+                - A string containing a http link pointing to an image
```
```suggestion
                - A string containing a HTTP(S) link pointing to an image
```

mishig25 committed:
Update src/transformers/pipelines/object_detection.py

Co-authored-by: NielsRogge <48327001+NielsRogge@users.noreply.github.com>

mishig25 committed:
Update src/transformers/pipelines/object_detection.py

Co-authored-by: NielsRogge <48327001+NielsRogge@users.noreply.github.com>

mishig25 committed:
Update src/transformers/pipelines/object_detection.py

Co-authored-by: NielsRogge <48327001+NielsRogge@users.noreply.github.com>

NielsRogge commented on src/transformers/pipelines/object_detection.py:
```diff
@@ -0,0 +1,178 @@
+                - A string containing a local path to an image
+                - An image loaded in PIL directly
+
+                The pipeline accepts either a single image or a batch of images, which must then be passed as a string.
```
The images shouldn't necessarily be passed as a string, so I would leave that away.

mishig25 replied:
Great catch!

NielsRogge commented on src/transformers/pipelines/object_detection.py:
```diff
@@ -0,0 +1,178 @@
+                - An image loaded in PIL directly
+
+                The pipeline accepts either a single image or a batch of images, which must then be passed as a string.
+                Images in a batch must all be in the same format: all as http links, all as local paths, or all as PIL
```
```suggestion
                Images in a batch must all be in the same format: all as HTTP(S) links, all as local paths, or all as PIL
```

mishig25 committed:
Update src/transformers/pipelines/object_detection.py

Co-authored-by: NielsRogge <48327001+NielsRogge@users.noreply.github.com>

mishig25 committed:
Update src/transformers/pipelines/object_detection.py

Co-authored-by: NielsRogge <48327001+NielsRogge@users.noreply.github.com>

NielsRogge commented on src/transformers/pipelines/object_detection.py:
```diff
@@ -0,0 +1,178 @@
+
+            The dictionaries contain the following keys:
+
+            - **label** (:obj:`str`) -- The label identified by the model.
```
```suggestion
            - **label** (:obj:`str`) -- The class label identified by the model.
```

mishig25 committed:
Update src/transformers/pipelines/object_detection.py

Co-authored-by: NielsRogge <48327001+NielsRogge@users.noreply.github.com>

mishig25 committed:
Rm timm requirement

mishig25 committed:
make fixup

mishig25 commented:
@LysandreJik I think its ready to be merged. Please let me know if you there's anything else I need to take care of :)

Narsil commented:
Hi @mishig25 ,

I think you need to fix all the tests.

`import torch` need to be protected behind `is_torch_available` for instance.
For code quality you can `pip install -e .[dev]` and then `make fixup`.
The PT tests also seem to require `timm` which are not available in the tests. So you need a `@require_timm` decorator.

mishig25 committed:
Merge branch 'master' into object_detection_pipeline

mishig25 commented:
@Narsil I'm confused about the tests failing.
For example, in this[ failed test](https://app.circleci.com/pipelines/github/huggingface/transformers/27659/workflows/e994e3b6-f627-477f-ba14-24bda195f91c/jobs/268944), I see the test failing for pipelines I haven't made any changes (also, I made sure my branch is up-to-date with the master):
here is an example for **test_pipelines_translation**
```bash
_____________ ERROR collecting tests/test_pipelines_translation.py _____________
ImportError while importing test module '/home/circleci/transformers/tests/test_pipelines_translation.py'.
...
E   ModuleNotFoundError: No module named 'torch'
```
Please let me know what step I'm missing

mishig25 committed:
Add timm requirement to test

mishig25 committed:
Make fixup

mishig25 committed:
Guard torch.Tensor

mishig25 committed:
Chore

mishig25 commented on src/transformers/pipelines/object_detection.py:
```diff
@@ -0,0 +1,177 @@
+
+        return annotations
+
+    def _get_bounding_box(self, box: "torch.Tensor") -> Dict[str, int]:
+        """
+        Turns list [xmin, xmax, ymin, ymax] into dict { "xmin": xmin, ... }
+
+        Args:
+            box (torch.Tensor): Tensor containing the coordinates in corners format.
+
+        Returns:
+            bbox (Dict[str, int]): Dict containing the coordinates in corners format.
+        """
+        if self.framework != "pt":
+            # guarding `box: "torch.Tensor"`
+            raise ValueError("The ObjectDetectionPipeline is only available in PyTorch.")
+        xmin, ymin, xmax, ymax = box.int().tolist()
```
@Narsil guarding "torch.Tensor" here. Please let me know what you think about it

mishig25 committed:
Delete unnecessary comment

mishig25 commented:
Since the PR was approved by two HF members and tests passed, I've merged it when the merge option became available. Please let me know if it is a correct procedure (i.e. should I have waited until a transfomers maintainer merged it?)

sgugger commented:
That's correct: as long as you have approval of one core maintainer (more for big PRs), addressed all comments, and all tests pass, you can merge your PR. :-)
